core/domain/question_domain.py

```diff
@@ -203,9 +203,39 @@ def _convert_state_v28_dict_to_v29_dict(cls, question_state_dict):
         question_state_dict['solicit_answer_details'] = False
         return question_state_dict
 
+    @classmethod
+    def _convert_state_v29_dict_to_v30_dict(
+            cls, question_state_dict, linked_skill_ids):
+        """Converts from version 29 to 30. Version 30 replaces
+        tagged_misconception_id with tagged_skill_misconception_id, which
+        contains the skill id and misconception id of the tagged misconception,
+        connected by '-'.
+
+        Args:
+            question_state_dict: dict. A dict where each key-value pair
+                represents respectively, a state name and a dict used to
+                initalize a State domain object.
+            linked_skill_ids: list(str). Skill ids linked to the question.
+
+        Returns:
+            dict. The converted question_state_dict.
+        """
+        answer_groups = question_state_dict['interaction']['answer_groups']
+        for answer_group in answer_groups:
+            if answer_group['tagged_misconception_id'] != None:
+                answer_group['tagged_skill_misconception_id'] = (
+                    linked_skill_ids[0] + '-' +
+                    str(answer_group['tagged_misconception_id']))
+            else:
+                answer_group['tagged_skill_misconception_id'] = None
+
+            del answer_group['tagged_misconception_id']
+        return question_state_dict
+
     @classmethod
     def update_state_from_model(
-            cls, versioned_question_state, current_state_schema_version):
+            cls, versioned_question_state, current_state_schema_version,
+            question_model):
```

Comment:
Why are we sending Models in the domain layer?

Reply:
As you can see in the function below, I need the linked_skill_ids list in the question_model to get the first linked skill id.

Reply:
Then pass the linked_skill_ids. Don't pass the entire model; the *_domain.py files should not be dealing with the model layer or any model instances at all.

Reply:
Actually, scratch that. I have a simpler solution. We decided to clear all the new structures models from production servers recently. So, for this PR, you can just replace everything with None (ignore any existing ids). Does that help?

Reply:
Done.